torch/_dynamo/symbolic_convert.py

```diff
@@ -581,6 +581,34 @@ def handle_graph_break(
     return decorator
 
 
+class BackedgeTracker:
+    """
+    A BackedgeTracker is used to keep track of backedges and how many nodes
+    we're expanding the graph by each time the backedge is seen. The general
+    idea is that if you're looping and each loop is inlining a large sub-graph
+    we detect that and skip frame when it's looking untenable.
+    """
+
+    def __init__(self):
+        self.n_seen = 0
+
+    # Raises SkipFrame if the loop is getting too big.
+    def append(self, count):
+        self.n_seen += 1
+
+        # Don't skip if we haven't seen this particular backedge at least a few
+        # times.
+        if self.n_seen < 3:
+            return
+
+        # For now use the trivial hueristic of checking the raw number of nodes
+        # in the graph. In the future we could do something more interesting
+        # like watching the rate of growth to trim loops earlier (so we don't
+        # have to wait for `max` nodes before skipping).
+        if config.max_loop_unroll_nodes > 0 and count > config.max_loop_unroll_nodes:
+            raise exc.SkipFrame("unrolled loop getting too big")
```

Comment:
Can you explain how this fixes the problem? When we throw this exception we just graph break and fall back to eager mode?

Reply:
That's right. This doesn't make the underlying problem go away - it just makes the pathological case not misbehave.

Reply:
I'm thinking maybe I should change this subtly so instead of checking for total nodes it checks for nodes added by this particular backedge (so 1 loop expanding into 6000 nodes is not expanded but 4 independent loops expanding into 2000 nodes each is fine). The point is to be a backstop - not a hard limit.

Reply:
Also maybe this should treat full_graph mode as '0' as well...

Reply:
> Can you explain how this fixes the problem? When we throw this exception we just graph break and fall back to eager mode?

@oulgen Actually this is not graph break, this is skip frame, which means the whole frame doesn't get compiled.